src/Razor/Microsoft.AspNetCore.Razor.Language/src/Extensions/ViewCssScopePass.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Extensions
+{
+    internal class ViewCssScopePass : IntermediateNodePassBase, IRazorOptimizationPass
+    {
+        // Runs after taghelpers are bound
+        public override int Order => 110;
+
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            var cssScope = codeDocument.GetCssScope();
+            if (string.IsNullOrEmpty(cssScope))
+            {
+                return;
+            }
+
+            if (!string.Equals(documentNode.DocumentKind, "mvc.1.0.view", StringComparison.Ordinal) &&
+                !string.Equals(documentNode.DocumentKind, "mvc.1.0.razor-page", StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            var nodes = documentNode.FindDescendantNodes<HtmlContentIntermediateNode>();
+            for (var i = 0; i < nodes.Count; i++)
+            {
+                ProcessElement(nodes[i], cssScope);
+            }
+        }
+
+        private void ProcessElement(HtmlContentIntermediateNode node, string cssScope)
+        {
+            cssScope = " " + cssScope;
+            // Add a minimized attribute whose name is simply the CSS scope
+            for (var i = 0; i < node.Children.Count; i++)
+            {
+                var child = node.Children[i];
+                if (child is IntermediateToken token && token.IsHtml)
+                {
+                    var content = token.Content;
+                    if (content.StartsWith("<") && !content.StartsWith("</"))
```

Comment:
These should use `StringComparison.Ordinal`.

Reply:
I don't think it matters, does it?

Reply:
Might be beneficial to avoid any present or future linter warnings. But agreed it shouldn't affect behavior.

I've been trying to think of any other cases where we might find a `<` which isn't immediately preceding a tag name, and haven't thought of any. For example cases like `< p >` or `<a href=< >...</a>` but I think these are too invalid to get through the parser anyway. So as far as I can tell, this seems fine.

Reply:
Yes, I had the same concern and experimented a bit with different syntaxes. Worst case scenario (unlikely)  we get a "smalish" bug here, that has an easy workaround and is easy to fix.